ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,294 @@
+-- dynamo, then it starts as a dynamo; otherwise, it starts as a jumper.
+registerClient ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  -- | A function to make a client handle from a jumping state.
+  (ChainSyncJumpingState m blk -> ChainSyncClientHandle m blk) ->
+  STM m ()
+registerClient handlesVar peer mkHandle = do
+  cschJumping <-
+    getDynamo handlesVar >>= \case
+      Nothing -> pure $ Dynamo Origin
+      Just _ -> newJumper GenesisPoint Happy
+  modifyTVar handlesVar $ Map.insert peer $ mkHandle cschJumping
+
+unregisterClient ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  STM m ()
+unregisterClient handlesVar peer = do
+  handle <- (Map.! peer) <$> readTVar handlesVar
+  modifyTVar handlesVar $ Map.delete peer
+  case cschJumping handle of
+    Jumper _ _ _ -> pure ()
+    Objector _   -> electNewObjector handlesVar
+    Dynamo _     -> electNewDynamo handlesVar
+
+-- | Look into all objector candidates and promote the one with the oldest
+-- intersection with the dynamo as the new objector.
+electNewObjector ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  STM m ()
+electNewObjector handlesVar = do
+  demoteObjector handlesVar
+  -- Get all the jumpers that disagree with the dynamo and for which we know the
+  -- precise intersection; find the one with the oldest intersection (which is
+  -- very important) and promote it to objector.
+  sortedObjectorCandidates <-
+    List.sortOn (pointSlot . snd)
+      . mapMaybe
+        ( \(peer, handle) ->
+            case cschJumping handle of
+              Jumper _ intersection FoundIntersection -> Just (peer, intersection)
+              _ -> Nothing
+        )
+      . Map.toList
+      <$> readTVar handlesVar
+  case sortedObjectorCandidates of
+    [] -> pure ()
+    (objector, intersection) : _ ->
+      setJumpingState handlesVar objector $ Objector intersection
+
+electNewDynamo ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  STM m ()
+electNewDynamo handlesVar = do
+  -- Get everybody; choose an unspecified new dynamo and put everyone else back
+  -- to being jumpers. NOTE: This might stop an objector from syncing and send
+  -- it back to being a jumper only for it to start again later, but nevermind.
+  (Map.keys <$> readTVar handlesVar)
+    >>= \case
+      [] -> pure ()
+      dynamo : jumpers -> do
+        setJumpingState handlesVar dynamo $ Dynamo Origin
+        forM_ jumpers $ \jumper ->
+          setJumpingState handlesVar jumper =<< newJumper GenesisPoint Happy
+
+-- | Size of jumps, in number of slots. FIXME: Make this configurable.
+jumpSize :: SlotNo
+jumpSize = 2
```

Comment:
Why is the jump so small?

Reply:
Just because I wanted to see it in peer simulator tests. This will actually need to come from a configuration option somewhere that will also allow disabling CSJ altogether.

In general, I don't think the actual value of jump sizes for main net is set. @nfrisby and @amesgen mentioned half the size of a Genesis window, but I don't think that's set in stone. Some of the considerations I remember are:

- The tinier the jumps, the heavier the work on the network (and the honest nodes in particular).

- The longer the jumps, the more imprecise the information is, which has several impacts:

  - In the current implementation, this means that more time is spent with back-and-forth `MsgFindIntersection` to pinpoint the exact intersection. (There is definitely work to do on making this process faster but it will probably always remain logarithmic in the size of the jumps.)

  - When an objector boots up, they might be `jumpSize` slots behind the dynamo, so they might take time to catch up (on purpose or just because that's many headers), and that's time spent not moving forward with the sync. This should happen very rarely in practice because that then eliminates an adversary from the list of peers.

  - If `jumpSize` is close to the size of the forecast window, the delay of response for the jump requests will start implying time spent with the dynamo beyond the forecast horizon, which is time spent not exchanging headers and potentially where the TCP window shrinks.

I'm sure there are many other considerations I don't remember but that's it off the top of my head.